src/Spelling/CodeAnalysisDictionary.cs

```diff
@@ -0,0 +1,207 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Xml;
+
+namespace Spelling
+{
+    /// <summary>
+    /// Parses and stores words/terms obtained from a custom dictionary.
+    /// </summary>
+    internal sealed class CodeAnalysisDictionary
+    {
+        public CodeAnalysisDictionary(DictionaryInfo dictionary)
+        {
+            RecognizedWords = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            UnrecognizedWords = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            CasingExceptions = new HashSet<string>();   // this _is_ case-sensitive
+            DiscreteExceptions = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            DeprecatedAlternateWords = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+            CompoundAlternateWords = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+
+            LoadWordsIfNecessary(dictionary);
+        }
+
+        /// <summary>
+        /// Latest date and time of the custom dictionary's last write time/creation time.
+        /// </summary>
+        public DateTime LastUpdated { get; private set; }
```

Comment:
You don't need such a custom caching infrastructure for storing custom data per-additional file. [AnalysisContext.TryGetValue](http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs,211) already provides this functionality. You just create your own [SourceTextValueProvider](http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/SourceTextValueProvider.cs,820d323c07c3f549) that takes the delegate to compute data for a given additional text, and a value comparer to compare additional text ([default implementation](http://source.roslyn.io/#Microsoft.CodeAnalysis/Text/SourceTextComparer.cs,b6556886ea9ba2b6,references) performs checksum comparison)

Reply:
First, thanks for taking the time to review this. Second, good to know. When I get a chance I'll look into this approach.

Reply:
No longer applies